### ISSUE TITLE
iterable.return() is not sufficient to clean up subscriptions

### ISSUE THREAD

**Issue by jquense** [apollographql/graphql-subscriptions#143]
Hey there 👋 This is perhaps a bug report and a bit of a question about how ya;'ll have implemented this. 

We've been working on our own semi-internal implementation of subscriptions for graphql and come across an issue handling the pub-sub case, one that seems like ya'll suffer from as well (unless i'm missing something)

The crux of the issue is that `iterable.return()` doesn't get called when an iterable hasn't started. which means once you've used language level constructs to map or filter an iterable it'll never clean up unless it's already had a value pulled

Following will run in Chrome if you paste it into a console:

```js
let f = (()=> {
  let i = 0
  return {
    next() {
      return Promise.resolve({ value: i++, done:  i >= 4 })
    },
    return() {
      console.log('return!')
      return Promise.resolve({ value: undefined, done: true });
    },
    throw(error) {
      return Promise.reject(error);
    },
    [Symbol.asyncIterator]() {
      return this
    }
  }
})()

async function* addOne() {
  for await (const val of f) yield val + 1;
}

await addOne().return();
```

The practical problem is that if you disconnect or unsubscribe from a subscription before it pushes a value, the underlying pup-sub will never clean itself up.

**Comment by jedwards1211** [comment]
I don't believe this is a problem for subscriptions Apollo because I don't think it uses `for await` anywhere under the hood.

Looking at `subscriptions-transport-ws`, I believe it calls an `AsyncIterator`'s `return` method here: https://github.com/apollographql/subscriptions-transport-ws/blob/master/src/server.ts#L211

But keep in mind: no `AsyncIterable`s anyone is using for GraphQL subscriptions should do anything that necessitates cleanup **until their `Symbol.asyncIterator` method is called**.  As long as they follow that contract then they will work safely in `for await` statements without leaving any dangling resources.

**Comment by jedwards1211** [comment]
However, this package's `eventEmitterToAsyncIterator` method does commit the sin of adding event listeners before its `Symbol.asyncIterator` method is called.  I think in the context of using it with `subscriptions-transport-ws` its `return` method will always be called, but still, it's not adding event listeners in the safest place in general, so I think @jquense was right to bring this up.

**Comment by taion** [comment]
See https://github.com/davidyaha/graphql-redis-subscriptions/blob/v1.4.0/src/pubsub-async-iterator.ts, for example, which, while not exactly first-party, also doesn't quite do the right thing.

**Comment by taion** [comment]
cc @davidyaha

**Comment by jedwards1211** [comment]
Actually this is a very important issue because if one writes a valid `concat` function for async iterables and then `concat`s two `pubsub.asyncIterator`s, one or both of those could leave dangling listeners after the top-level async iterable has been `return`ed.

**Comment by jedwards1211** [comment]
`concat` is maybe a bad example but I've been meaning to write a utility function for prepending the current value at the start of a subscription (since there can be a window of time in between the query result and first subscription result during which updates get lost 😱 )

**Comment by jedwards1211** [comment]
Additionally, `withFilter` also suffers from this danger and needs to be redesigned to operate on `AsyncIterable`s rather than `AsyncIterator`s themselves.

**Comment by jedwards1211** [comment]
@jquense I think the title of this issue isn't really correct, `return()` is perfectly sufficient to clean up subscriptions, the async iterables just have to be written correctly.

**Comment by jquense** [comment]
Thanks for taking look! I'm not sure tho how the iterable/iterator distinction helps the the above example with the generator does return an iterable, main issue is return() doesn't do anything unless the first yield has been hit. Passing the iterable to a valid map or filter method will still break the chain unless I'm misunderstanding what your proposing? Thanks again for jumping on this !

**Comment by jquense** [comment]
Sorry by "above examples" I meant if you used an async generator you'd hit the same problem

**Comment by jedwards1211** [comment]
@jquense Okay let me clarify, cause yeah this stuff is not obvious.

In `for await (let val of foo)`, `foo` must be an `AsyncIterable`, but not necessarily `AsyncIterator`.  When the loop starts it calls `foo[Symbol.asyncIterator]` to create the actual `AsyncIterator`, which can be a completely different object than `foo` (and ought to be a different object in most use cases).  The loop then repeatedly awaits `next()` on the iterator until it is done, in which case it calls the iterator's `return()` method, or an error is thrown, in which case it calls the iterator's `throw()` method.

If the `for await` loop never runs (like in your example) it never calls the `[Symbol.asyncIterator]` method.  Which means if you only perform listener registration when `[Symbol.asyncIterator]()` runs, you won't register listeners unless the loop is actually going to run (in which case you're guaranteed to get a `return()` or `throw()` call in which you can unregister the listeners).

So here's a safe way to rewrite your original example:

```js

const iterable = {
  [Symbol.asyncIterator]() {
    let i = 0
    console.log('adding listener!')
    // below is the actual iterator -- this creates a new iterator with each call
    return {
      next() {
        return Promise.resolve({ value: i++, done:  i >= 4 })
      },
      return() {
        console.log('removing listener!')
        return Promise.resolve({ value: undefined, done: true });
      },
      throw(error) {
        console.log('removing listener!')
        return Promise.reject(error);
      },
    }
  }
}

async function* addOne(iterable) {
  for await (const val of iterable) yield val + 1;
}
```
Now if I call the generator but make its iterator `return()` before I ever call `next()`, it never registers the "listener":
```
> let iterator = addOne(iterable)
undefined

> await iterator.return()
{value: undefined, done: true}
```
However, the first time I call `next()`, it actually runs the `for await` statement, and the first thing that does is call `iterable[Symbol.asyncIterator]()`, which is what actually adds the listener.  And when I call `return()` in the middle of the `for await`, it does pass through the return call, removing the listener:
```
> iterator = addOne(iterable)
addOne {<suspended>}

> await iterator.next()
adding listener!
{value: 1, done: false}

> await iterator.return()
removing listener!
{value: undefined, done: true}
```

**Comment by jquense** [comment]
hmm i'm surprised that suffer the same problem but if next() doesn't resolve before return is called? The situation were we ran into this wasn't that next was never called, it was immediately, but until the first event was received, return didn't get called through. The other downside here is that you not can't use syntax to create the iterable, since you need to customize the iterator method, but maybe that's ok since it's just the consumer

**Comment by jquense** [comment]
which is to say this is still a failure condition:
```js
const iterable = {
  [Symbol.asyncIterator]() {
    let i = 0
    console.log('adding listener!')
    return {
      next() {
        return Promise.resolve({ value: i++, done:  i >= 4 })
      },
      return() {
        console.log('removing listener!')
        return Promise.resolve({ value: undefined, done: true });
      },
      throw(error) {
        console.log('removing listener!')
        return Promise.reject(error);
      },
    }
  }
}

const iter = iterable[Symbol.asyncIterator]]() // adding listeners

iterable.return() // nothing
```

**Comment by jquense** [comment]
huh maybe i've been over complicating this. It's pretty straightforward to handle this with a generator.

```js
var addListeners = () => console.log('adding listener!');
async function* create() {
  let i = 0;
  try {
    await addListeners();

    while (i < 5) yield ++i;
  } finally {
    console.log('remove listeners');
  }
}

async function* addOne(iterable) {
  for await (const val of iterable) yield val + 1;
}

var iter = addOne(create());

iter.return();
```
perhaps rather than doing the logic in the iterator method, its best to move it to inside the first next invocation?

**Comment by jedwards1211** [comment]
@jquense that should be:

```js
const iterator = iterable[Symbol.asyncIterator]() // adding listeners

iterator.return() // removing listeners
```
No problem in this case.

**Comment by jedwards1211** [comment]
> if `next()` doesn't resolve before `return()` is called

I could be wrong about this, but I think when using async iterators you basically are responsible for calling `return` after all `next` calls are resolved.

In any case adding the listeners when the iterator is created is less vulnerable to an early `return` call than adding the listeners in the first call to `next`.

**Comment by jedwards1211** [comment]
Is it possible to yield events coming from a pubsub system using a generator though?  Seems like it should, I just can't think of it at the moment...

**Comment by jquense** [comment]
> No problem in this case.

It doesn't work that way tho? 

> Is it possible to yield events coming from a pubsub system using a generator though?

Yeah it is but with some annoyance, we use a queue abstraction: https://github.com/4Catalyzer/graphql-subscription-server/blob/master/src/AsyncUtils.js#L31

**Comment by jedwards1211** [comment]
In any case the thing graphql-subscriptions does wrong is, it doesn't currently provide a method to create an iterable that waits to add the listeners until the iterable actually gets used.  Which is what I added in my PR.

**Comment by jedwards1211** [comment]
```
> const iterable = {
  [Symbol.asyncIterator]() {
    let i = 0
    console.log('adding listener!')
    return {
      next() {
        return Promise.resolve({ value: i++, done:  i >= 4 })
      },
      return() {
        console.log('removing listener!')
        return Promise.resolve({ value: undefined, done: true });
      },
      throw(error) {
        console.log('removing listener!')
        return Promise.reject(error);
      },
    }
  }
}
undefined

> var iterator = iterable[Symbol.asyncIterator]()
VM59:4 adding listener!
undefined

> await iterator.return()
VM59:10 removing listener!
{value: undefined, done: true}
```

**Comment by jedwards1211** [comment]
It may be okay in this lib to move the subscription inside the first `next()` invocation, but:
* It's unnecessarily wasteful to check that with each `next` invocation
* `[Symbol.asyncIterator]() { return this }` is kind of bad, it violates the expectation that the method returns a fresh new iterator every time.

**Comment by jquense** [comment]
> [Symbol.asyncIterator]() { return this } is kind of bad, it violates the expectation that the method returns a fresh new iterator every time.

I don't think that's the correct expectation though. native behavior doesn't return a fresh iterator:

```js
async function* create() {
  let i = 0;
  while (i < 5) yield ++i;
}
var iterable = create()

iterable[Symbol.asyncIterator]() === iterable[Symbol.asyncIterator]() // true

await iterable[Symbol.asyncIterator]().next() // 1
await iterable[Symbol.asyncIterator]().next() // 2
```

**Comment by jedwards1211** [comment]
@jquense I see.

As far as I understand generator functions are mainly intended to be called every time you want to create an iterator, and calling `[Symbol.asyncIterator]` multiple times on its return value is not really intended.  I guess they made the created iterator return itself in its `Symbol.asyncIterator` method so that it would be more convenient to do things like `for await (let val of create())`.

**Comment by taion** [comment]
It's worth keeping in mind FWIW that the issue isn't generator functions per se. It's about adapting external event sources (something like an observable) to an async iterable.

Right now these options all seem pretty mediocre.

**Comment by jedwards1211** [comment]
@taion yes, it is a bit tricky.  I do think GraphQL was wise to use async iterables because they are the first built-in feature added to JS that is equivalent to observables.  What are your thoughts on my PR?

**Comment by jedwards1211** [comment]
I think using a generator that adds listeners before it yields anything (like @jquense suggested) would be a pretty decent approach.  Only thing I don't know is, does TypeScript support async generators?  If not I could just move the listener registration inside the first call to `next()` in `eventEmitterAsyncIterator`.

**Comment by jquense** [comment]
this is where we landed btw: https://github.com/4Catalyzer/graphql-subscription-server/blob/master/src/AsyncUtils.js#L55 it works well (so far), but yeah it's a bit easier to do with syntax than a hand made one, tho that could be wrapped up in a helper pretty easily

**Comment by taion** [comment]
From the other thread, this is much worse than previously understood given https://github.com/tc39/proposal-async-iteration/issues/126#issuecomment-403454433.

At the current point, `iterator.return()` is just not going to work for tearing down and cleaning up resources.

**Comment by jedwards1211** [comment]
@taion I understand you are frustrated and would like to see a sea change here, but please be technically honest: `iterator.return()` will work as long as people don't use async generators.

I think it's important to be clear about this because solving this issue by switching to observables will be a far-reaching change requiring coordination between many different parts of the ecosystem, not necessarily something I'm opposed to, but if it even happens it will take a lot of time.  So until then, it's important that we spread the word that people should avoid using async generators for GraphQL subscriptions and try to clearly explain why.

**Comment by jquense** [comment]
I think while true the caveat that "people not use async generators" is extreme and deal-breaking enough to warrant "not going to work", when the syntax designed to create these iterators actively prevents correct behavior. This almost certainly precludes using iteration libraries and lodash like utils, since they will likely use the syntax. The main ergonomic benefit here is the syntax to begin with, hand spun iterators aren't particularly compelling as an API.  that's reason enough for me to think that there should be a different API. I know we can't necessarily control that but it also doesn't seem valuable to keep trying to shoehorn the usage of an language feature that is not designed to handle the use-case.

IMO graphql-js should back-out this api and pick something with the correct affordances (not that anyone here can make that choice)

**Comment by TimSusa** [comment]
Hi, 

as contributor of https://github.com/axelspringer/graphql-google-pubsub I can say we run into big trouble with our current implementation, causing memory leaks on production. See that screenshot:

![screen shot 2018-08-22 at 11 33 41](https://user-images.githubusercontent.com/8963529/44478980-8b98d680-a63f-11e8-86f3-1839edc28637.png)

AsyncIterator seems to be the bad guy in the game.

Just today we discovered the issue and we ask ourselves, what is the right way to do it better with that async-beast-iter-something? I mean the hole disscussion seems to be not clear to us. It would be kind if someone could sumarize or provide us with a full example.

Anyway, we suspect, this PR could help: https://github.com/apollographql/graphql-subscriptions/pull/147 to realize https://github.com/davidyaha/graphql-redis-subscriptions/pull/114 , which maybe would help us at: https://github.com/axelspringer/graphql-google-pubsub to do it that way, but this is only a wild guess.

Anyway, any help would be appreciated.

Thanks in advance.

**Comment by taion** [comment]
@TimSusa This is how we ended up handling things, BTW: https://github.com/4Catalyzer/graphql-subscription-server/pull/20. We still expose an async iterable, because they're nice to work with for various operations, but we also thread through a separate "close" method that we use to clean things up.

**Comment by loudmouth** [comment]
@jedwards1211 i've been scouring the net trying to understand how exactly the `throw` function of the iterator is actually invoked. your comment is the first i've seen that conceptualizes the control flow for how `throw` is invoked, but in practical terms, i never see it executed.

my use case is that an end user can open a new subscription over an already existing websocket connection, but that new subscription may be rejected by the backend as a result of an expired token. i'd like to propagate that error back to the client, which is responsible for closing/reopening the websocket connection in this scenario.

any tips?

**Comment by jedwards1211** [comment]
@loudmouth Actually it's kind of irrelevant.  As it turns out, `graphql-js` doesn't require the `subscribe` method to return an `AsyncIterable`; it can instead return a `Promise<AsyncIterable>`.  So your subscribe method can be something like:

```js
const resolvers = {
  Subscription: {
     Foo: {
       async subscribe(src, args, context) {
         if (isTokenExpired(context)) throw new UnauthorizedError('token expired')
         return context.pubsub.asyncIterator('foo')
       }
     }
  }
}
```
The other option would be to make your async iterator's `next()` method return a `Promise` that rejects with the error you want (I assume `graphql-js` also supports this, but I haven't tested it).  Whether `throw()` gets called on your async iterator has no effect on communicating an error to GraphQL.

**Comment by ciekawy** [comment]
probably a bit off topic yet I got here looking for how to iterate my own in a standard way over `pubsub.asyncIterator` and got really surprised I cannot just do `for async` or apply RxJS. Or maybe I'm missing something?

**Comment by jedwards1211** [comment]
You can iterate with this syntax (inside of an async function):
```js
for await (const element of myAsyncIterable) {
}
```

**Comment by ciekawy** [comment]
```typescript
for await (const msg of pubSub.asyncIterator('foo')) {
}  
```

gives:

```
Type 'AsyncIterator<unknown, any, undefined>' must have a '[Symbol.asyncIterator]()' method that returns an async iterator.ts(2504)
```

**Comment by jedwards1211** [comment]
Sigh, most of the pubsub implementations return an object that's both an iterator and iterable (which is idiotic because it leads to confusion like this).  The method name is bad because you can't for await an iterator unless it also implements the iterable interface.  The correct return type for the method is AsyncIterableIterator.  I hate that things are done this way in JS because there was a clear distinction between iterators and iterables in Java.
